FV3/atmos_cubed_sphere/tools/coarse_grained_diagnostics.F90

```diff
@@ -72,127 +282,418 @@ subroutine initialize_coarse_diagnostic_axes(coarse_domain, &
+        trim(coarse_diag%name)
+      call mpp_error(FATAL, error_message)
+    endif
+   end subroutine coarse_grain_3D_field_on_model_levels
+
+   subroutine coarse_grain_3D_field_on_pressure_levels(is, ie, js, je, is_coarse, ie_coarse, js_coarse, je_coarse, &
+                                                       npz, coarse_diag, masked_area, masked_mass, phalf, upsampled_coarse_phalf, &
+                                                       ptop, result)
+    integer, intent(in) :: is, ie, js, je, is_coarse, ie_coarse, js_coarse, je_coarse, npz
+    type(coarse_diag_type) :: coarse_diag
+    real, intent(in) :: masked_mass(is:ie,js:je,1:npz), masked_area(is:ie,js:je,1:npz)
+    real, intent(in) :: phalf(is:ie,js:je,1:npz+1), upsampled_coarse_phalf(is:ie,js:je,1:npz+1)
+    real, intent(in) :: ptop
+    real, intent(out) :: result(is_coarse:ie_coarse,js_coarse:je_coarse,1:npz)
+
+    real, allocatable :: remapped_field(:,:,:)
+    character(len=256) :: error_message
+
+    allocate(remapped_field(is:ie,js:je,1:npz))
+
+    call vertically_remap_field( &
+      phalf, &
+      coarse_diag%data%var3, &
+      upsampled_coarse_phalf, &
+      ptop, &
+      remapped_field)
+    if (trim(coarse_diag%reduction_method) .eq. AREA_WEIGHTED) then
+      call weighted_block_average( &
+        masked_area(is:ie,js:je,1:npz), &
+        remapped_field(is:ie,js:je,1:npz), &
+        result &
+      )
+    elseif (trim(coarse_diag%reduction_method) .eq. MASS_WEIGHTED) then
+      call weighted_block_average( &
+        masked_mass(is:ie,js:je,1:npz), &
+        remapped_field(is:ie,js:je,1:npz), &
+        result &
+      )
+    else
+      write(error_message, *) 'coarse_grain_3D_field_on_pressure_levels: invalid reduction_method, ' // &
+        trim(coarse_diag%reduction_method) // ', provided for 3D variable, ' // &
+        trim(coarse_diag%name)
+      call mpp_error(FATAL, error_message)
+    endif
+   end subroutine coarse_grain_3D_field_on_pressure_levels
+
+   subroutine coarse_grain_2D_field(is, ie, js, je, npz, is_coarse, ie_coarse, js_coarse, je_coarse, &
+                                    Atm, coarse_diag, height_on_interfaces, result)
+    integer, intent(in) :: is, ie, js, je, npz, is_coarse, ie_coarse, js_coarse, je_coarse
+    type(fv_atmos_type), intent(in) :: Atm
+    type(coarse_diag_type), intent(in) :: coarse_diag
+    real, intent(in) :: height_on_interfaces(is:ie,js:je,1:npz+1)
+    real, intent(out) :: result(is_coarse:ie_coarse,js_coarse:je_coarse)
+
+    real :: log_pressure_level
+    character(len=256) :: error_message
+    real, allocatable :: work_2d(:,:), masked_area_weights(:,:)
+
+    if (coarse_diag%pressure_level > 0) then 
+      log_pressure_level = log(100.0 * real(coarse_diag%pressure_level))
+      allocate(work_2d(is:ie,js:je))
+      allocate(masked_area_weights(is:ie,js:je))
+      call mask_area_weights_single_pressure(Atm%gridstruct%area(is:ie,js:je), Atm%peln(is:ie,npz+1,js:je), &
```

Comment:
Note Lucas requests that we do not mask the values below the surface for these diagnostics.  I will update this when I get a chance.